src/compiler/factory.ts

```diff
@@ -1240,16 +1240,30 @@ namespace ts {
                     return binaryOperatorAssociativity === Associativity.Right;
                 }
                 else {
-                    // No need to parenthesize the right operand when the binary operator and
-                    // operand are the same and one of the following:
-                    //  x*(a*b)     => x*a*b
-                    //  x|(a|b)     => x|a|b
-                    //  x&(a&b)     => x&a&b
-                    //  x^(a^b)     => x^a^b
                     if (isBinaryExpression(operand)
-                        && operand.operatorToken.kind === binaryOperator
-                        && isMathAssociativeOperator(binaryOperator)) {
-                        return false;
+                        && operand.operatorToken.kind === binaryOperator) {
+                        // No need to parenthesize the right operand when the binary operator and
+                        // operand are the same and one of the following:
+                        //  x*(a*b)     => x*a*b
+                        //  x|(a|b)     => x|a|b
+                        //  x&(a&b)     => x&a&b
+                        //  x^(a^b)     => x^a^b
+                        if (isMathAssociativeOperator(binaryOperator)) {

```

Comment:
Maybe a better name is `operatorHasAssociativeProperty`, since that's usually the terminology used to describe it: https://en.wikipedia.org/wiki/Associative_property